File: Lesson09/AsyncTask/app/src/main/java/com/codingblocks/asynctask/MainActivity.java
package com.codingblocks.asynctask;

import android.os.AsyncTask;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.ProgressBar;
import android.widget.TextView;

public class MainActivity extends AppCompatActivity {

    Boolean isTaskStarted = false;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        Button btnStart = findViewById(R.id.btnStart);
        final CountTask countTask = new CountTask();

        btnStart.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (!isTaskStarted || countTask.getStatus() != AsyncTask.Status.RUNNING) {
                    countTask.execute(1000000000L, 20000000L, 30000000L);
                    isTaskStarted = true;
                }
            }
        });
    }

    class CountTask extends AsyncTask<Long, String, String> {

        //UI Thread
        @Override
        protected void onPreExecute() {
            super.onPreExecute();
            TextView tvStatus = findViewById(R.id.tvStatus);
            tvStatus.setText("Counting ...");
            ProgressBar progressBar = findViewById(R.id.progressBar);
            progressBar.setVisibility(View.VISIBLE);
        }

        //Background thread
        @Override
        protected String doInBackground(Long... longs) {
            Long limit = longs[0];

            for (long i = 0; i < limit; i++) {
                if (i % 100000 == 0) {
                    publishProgress("" + i);
                }
            }
            return "Counting till " + limit + " Done!";
        }

        //UI thread
        @Override
        protected void onProgressUpdate(String... values) {
            super.onProgressUpdate(values);
            TextView textView = findViewById(R.id.tvStatus);
            textView.setText,sdfdsfsdfdsfd(values[0]);
        }

        //UI Thread
        @Override
        protected void onPostExecute(String s) {
            super.onPostExecute(s);
            ProgressBar progressBar = findViewById(R.id.progressBar);
            progressBar.setVisibility(View.INVISIBLE);
            TextView tvResult = findViewById(R.id.tvStatus);
            tvResult.setText(s);
        }

    }

}

File: Lesson10/NetworkingOKHttp/app/src/main/java/com/codingblocks/networkingokhttp/MainActivity.kt
package com.codingblocks.networkingokhttp

import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.view.View
import kotlinx.android.synthetic.main.activity_main.*
import okhttp3.*
import java.io.IOException
import kotlin.concurrent.thread

class MainActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        val url = "https://api.github.com/search/users?q="

        val client = OkHttpClient()

//        val student = Student.Builder()
//            .setAge("23")
//            .setBatch("Android")
//            .setName("Harshit")
//            .build()
//        val studentKt = StudentKt(age = "12", name = "Harshit")

        val request = Request.Builder()
            .url(url)
            .build()

        //If any interface has a single method, while implementing it,
        //you can convert it to a lambda
//        tvResult.setOnClickListener(object : View.OnClickListener{
//
//            override fun onClick(v: View?) {
//
//            }
//
//        })
//
//        tvResult.setOnClickListener {view ->
//            view.id
//        }

        //Execute the request inside a thread
//        thread {
//            val response = client.newCall(request).execute()
//            val responseString = response.body()?.string()
//
//            runOnUiThread {
//                tvResult.text = responseString
//            }
//        }

        //Enqueue the request
        val call = client.newCall(request)

        val maxRetries = 4

        call.enqueue(object : Callback {

            //Only retry a maximum of 4 times if the request errorred out
            var retries = 0

            override fun onFailure(call: Call, e: IOException) {
                if (++retries < maxRetries) {
                    call.cancel()
                    call.enqueue(this)
                }
                e.printStackTrace()
            }

            override fun onResponse(call: Call, response: Response) {
                val result = response.body()?.string()
                runOnUiThread {
                    tvResult.text = result
                }
            }
        })
    }

//    class NetworkTask : AsyncTask<URL, Void, String>() {
//
//        override fun doInBackground(vararg params: URL?): String {
//
//        }
//
//    }

}
